File: solutions/2415.reverse-odd-levels-of-binary-tree.c
#include <stdbool.h>
#include <stdlib.h>

struct TreeNode {
  int val;
  struct TreeNode* left;
  struct TreeNode* right;
};

// @leet start
#define SIZE 16384

struct TreeNode* pop(struct TreeNode** stack, int* top) {
  if (*top != -1) {
    return stack[(*top)--];
  }

  return NULL;
}

void push(struct TreeNode** stack, int* top, struct TreeNode* val) { stack[++(*top)] = val; }

struct TreeNode** create() { return (struct TreeNode**)malloc(SIZE * sizeof(struct TreeNode*)); }

bool isEmpty(struct TreeNode** stack, int* top) { return (*top) == -1; }

struct TreeNode* reverseOddLevels(struct TreeNode* root) {
  struct TreeNode** stack = create();
  int top = -1;

  push(stack, &top, root);

  int level = 0;
  int current = 1;
  int left = current;

  while (!isEmpty(stack, &top)) {
    struct TreeNode* node = pop(stack, &top);
    left--;

    if (node != NULL) {
      push(stack, &top, node->left);
      push(stack, &top, node->right);

      node->val = level;
      // if (level % 2 == 1) {
      //   node->val = 0;
      // }

      if (left == 0) {
        current *= 2;
        left = current;
        level++;
      }
    }
  }

  return root;
}
// @leet end

File: solutions/2415.reverse-odd-levels-of-binary-tree.py
from types import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @leet start


class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        stack = []
        stack.append((root, 0))

        nodes = {}
        while stack:
            node, depth = stack.pop()

            if node is not None:
                if not nodes.get(depth):
                    nodes[depth] = []

                nodes[depth].append(node)

                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))

        for depth, line in nodes.items():
            if depth % 2 == 1:
                values = [node.val for node in line]
                for i in range(len(line)):
                    line[i].val = values[len(line) - i - 1]

        return root


# @leet end
